File: frontend/legacy/v1-archive-20260207/src/engine/backup/format.ts
import type { DetectResult } from '../ir/types';
import type { ArchiveEntries } from './archive';
import { hasFile, hasPrefix } from './archive';

export function detectFormat(entries: ArchiveEntries): DetectResult {
  const hints: string[] = [];
  const hasDataJson = hasFile(entries, 'data.json');
  const hasDataDir = hasPrefix(entries, 'Data');
  const hasSettingsJson = hasFile(entries, 'settings.json');
  const hasRikkaDb = hasFile(entries, 'rikka_hub.db');
  const hasUploadDir = hasPrefix(entries, 'upload');

  if (hasDataJson) hints.push('data.json');
  if (hasDataDir) hints.push('Data/');
  if (hasSettingsJson) hints.push('settings.json');
  if (hasRikkaDb) hints.push('rikka_hub.db');
  if (hasUploadDir) hints.push('upload/');

  if (hasDataJson && hasDataDir) {
    return { format: 'cherry', hints };
  }
  if (hasSettingsJson && (hasRikkaDb || hasUploadDir)) {
    return { format: 'rikka', hints };
  }
  return { format: 'unknown', hints };
}

File: internal/backup/format.go
package backup

import (
	"os"
	"path/filepath"
)

type Format string

const (
	FormatUnknown Format = "unknown"
	FormatCherry  Format = "cherry"
	FormatRikka   Format = "rikka"
)

type DetectResult struct {
	Format Format
	Hints  []string
}

func DetectExtractedDir(dir string) DetectResult {
	hints := make([]string, 0, 8)
	hasDataJSON := fileExists(filepath.Join(dir, "data.json"))
	hasDataDir := dirExists(filepath.Join(dir, "Data"))
	hasSettingsJSON := fileExists(filepath.Join(dir, "settings.json"))
	hasRikkaDB := fileExists(filepath.Join(dir, "rikka_hub.db"))
	hasUploadDir := dirExists(filepath.Join(dir, "upload"))

	if hasDataJSON {
		hints = append(hints, "data.json")
	}
	if hasDataDir {
		hints = append(hints, "Data/")
	}
	if hasSettingsJSON {
		hints = append(hints, "settings.json")
	}
	if hasRikkaDB {
		hints = append(hints, "rikka_hub.db")
	}
	if hasUploadDir {
		hints = append(hints, "upload/")
	}

	if hasDataJSON && hasDataDir {
		return DetectResult{Format: FormatCherry, Hints: hints}
	}
	if hasSettingsJSON && (hasRikkaDB || hasUploadDir) {
		return DetectResult{Format: FormatRikka, Hints: hints}
	}
	return DetectResult{Format: FormatUnknown, Hints: hints}
}

func fileExists(path string) bool {
	st, err := os.Stat(path)
	if err != nil {
		return false
	}
	return !st.IsDir()
}

func dirExists(path string) bool {
	st, err := os.Stat(path)
	if err != nil {
		return false
	}
	return st.IsDir()
}
